modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/components/A11y/A11y.tsx

```diff
@@ -0,0 +1,193 @@
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import './A11y.scss';
+
+import ClayIcon from '@clayui/icon';
+import ClayList from '@clayui/list';
+import ClayPopover from '@clayui/popover';
+import React, {useCallback, useEffect, useRef, useState} from 'react';
+import ReactDOM from 'react-dom';
+
+import useA11y from '../../hooks/useA11y';
+
+import type {Violation as TViolation} from '../../hooks/useA11y';
+import type {A11yCheckerOptions} from './A11yChecker';
+
+const rectAttrs: Array<keyof DOMRect> = [
+	'bottom',
+	'height',
+	'left',
+	'right',
+	'top',
+	'width',
+];
+
+interface IRectState {
+	rect: DOMRect | undefined;
+	hasRectChanged: boolean;
+}
+
+const rectChanged = (a: DOMRect = {} as DOMRect, b: DOMRect = {} as DOMRect) =>
+	rectAttrs.some((prop) => a[prop] !== b[prop]);
+
+const useObserveRect = (
+	callback: (rect: DOMRect | undefined) => void,
+	node: Element | null
+) => {
+	const rafIdRef = useRef<number>();
+
+	const run = useCallback(
+		(node: Element, state: IRectState) => {
+			const newRect = node.getBoundingClientRect();
+
+			if (rectChanged(newRect, state.rect)) {
+				state.rect = newRect;
+
+				callback(state.rect);
+			}
+
+			rafIdRef.current = window.requestAnimationFrame(() =>
+				run(node, state)
+			);
+		},
+		[callback]
+	);
+
+	useEffect(() => {
+		if (node) {
+			run(node, {
+				hasRectChanged: false,
+				rect: undefined,
+			});
+
+			return () => {
+				if (rafIdRef.current) {
+					cancelAnimationFrame(rafIdRef.current);
+				}
+			};
+		}
+	}, [node, run]);
+};
+
+const Overlay = React.forwardRef<
+	HTMLDivElement,
+	React.ButtonHTMLAttributes<HTMLDivElement>
+>(({style, ...othersProps}, ref) =>
+	ReactDOM.createPortal(
+		<div {...othersProps} className="a11y-overlay" ref={ref} style={style}>
+			<div className="a11y-indicator">
+				<ClayIcon symbol="info-circle" />
+			</div>
+			<div className="a11y-backdrop" />
+		</div>,
+		document.body
+	)
+);
+
+type ViolationProps = {
+	modifyIndex: number;
+	target: string;
+	violations: Array<TViolation>;
+};
+
+function Violation({target, violations}: ViolationProps) {
+	const [visible, setVisible] = useState(false);
+	const [bounds, setBounds] = useState<React.CSSProperties>();
+
+	const node = document.querySelector(target);
+
+	useObserveRect(
+		useCallback(
+			(bounds) => {
+				bounds = bounds ?? (node as Element).getBoundingClientRect();
+
+				setBounds({
+					height: bounds.height,
+					left: bounds.left,
+					top: bounds.top,
+					width: bounds.width,
+				});
+			},
+			[node]
+		),
+		node
+	);
+
+	return (
+		<ClayPopover
+			className="a11y-popover"
+			header={
+				<>
+					<div className="inline-item">
+						<ClayIcon
+							className="text-danger"
+							symbol="info-circle"
+						/>
+					</div>
+					<div className="inline-item inline-item-after">
+						<span>
+							{/* @ts-ignore */}
```

Comment:
Are we ignoring this because of the call to the global `Liferay` object?

Reply:
Yeah, I added a Liferay type declaration now (https://github.com/liferay-frontend/liferay-portal/pull/1044/commits/ce44638e19f125d03d3209304dd0b01691a729ac), I think we should start finding somewhere to add all Liferay types so that we can use them in projects.